R/MsBackendDataFrame.R

```diff
@@ -248,14 +248,31 @@ setReplaceMethod("smoothed", "MsBackendMemory", function(object, value) {
 })
 
 #' @rdname hidden_aliases
-setMethod("spectraData", "MsBackendMemory", function(object, columns) {
-    ## TODO @jo IMPLEMENT:
-    ## allow to specify columns. Have then to check if these columns exist,
-    ## if not call the respective method.
+#'
+#' @importFrom methods as
+setMethod("spectraData", "MsBackendDataFrame", function(object, columns) {
+    if (missing(columns))
+        columns <- union(colnames(object@spectraData),
+                         names(.SPECTRA_DATA_COLUMNS))
+    df_columns <- columns[columns %in% colnames(object@spectraData)]
+    res <- object@spectraData[, df_columns, drop = FALSE]
+    other_columns <- columns[!(columns %in% colnames(object@spectraData))]
+    if (length(other_columns)) {
+        other_res <- lapply(other_columns, .get_spectra_data_column, x = object)
+        names(other_res) <- other_columns
+        is_mz_int <- names(other_res) %in% c("mz", "intensity")
+        if (!all(is_mz_int))
+            res <- cbind(res, as(other_res[!is_mz_int], "DataFrame"))
+        if (any(names(other_res) == "mz"))
+            res$mz <- if (length(other_res$mz)) other_res$mz else list()
+        if (any(names(other_res) == "intensity"))
+            res$intensity <- if (length(other_res$intensity)) other_res$intensity else list()
+    }
+    res[, columns, drop = FALSE]
 })
 
 #' @rdname hidden_aliases
-setReplaceMethod("spectraData", "MsBackendMemory", function(object, value) {
+setReplaceMethod("spectraData", "MsBackendDataFrame", function(object, value) {
```

Comment:
Don't we need to check for column names here, making sure the mandatory ones are present?

Reply:
That's what I don't want to have - I would like to keep the `@spectraData` flexible and that not all *mandatory* columns should be present. I would also not call them mandatory - it's rather columns that have a mandatory data type and columns/variables that each backend needs to provide - whether or not the backend stores them is up to the backend. My `MsBackendDataFrame` does not store columns that are purely `NA`.

Reply:
Ok, I see. So rather than having these columns filled with `NA`, you prefer to generate them on the fly every time these data need to be accessed. 

An option would be to create them if absent, and simplify all the accessors to return these columns. I need to think about this a bit more. Or we can discuss this.

Reply:
Honestly, also looking at the size objects can get (I have a `OnDiskMSnExp` of *only* 1200 files where just the `fData` is 1.4GB large) I'd prefer the way to not require to always have all of these columns in the `@spectraData`.

Reply:
And could this be encoded as an `Rle`?
```r
> x <- rep(NA, 1e6)
> pryr::object_size(x)
4 MB
> pryr::object_size(Rle(x))
1.18 kB
```
I am thinking in terms of code (simpler accessors) and overhead of generating these long vectors repeatedly.

Reply:
Interestingly:
```r
> x <- rep(NA, 1e6)
> df1 <- DataFrame(x = x, y = x, z = x)
> df2 <- DataFrame(x = Rle(x), y = Rle(x), z = Rle(x))
> df3 <- DataFrame(x = rep(NA, 1e6), y = rep(NA, 1e6), z = rep(NA, 1e6))
> object_size(x)
4 MB
> object_size(Rle(x))
1.18 kB
> object_size(df1)
4 MB
> object_size(df2)
3.73 kB
> object_size(df3)
12 MB
```
Even without the `Rle`, if we use the same vector of `NA` values to the different variables, we don't multiply memory usage. This could even be reduced by reusing the Rle
```r
> xrle <- Rle(x)
> df4 <- DataFrame(x = xrle, y = xrle, z = xrle)
> object_size(df4)
2.74 kB
```

Reply:
I have to think over that. But also with the `Rle` there will be some more code necessary. We would have to un-rle if the column is an Rle in the accessors (which should be less of a problem) and also in the `spectraData` accessor. 

And I did not yet find a possibility to check the data type of an `Rle` - do you know how that could be done? `is.integer(Rle(NA_integer, 5))` is not working.

Reply:
```r
> x <- Rle(rep(NA_integer_, 1e3))
> str(as.integer(x))
 int [1:1000] NA NA NA NA NA NA NA NA NA NA ...
> str(as.vector(x))
 int [1:1000] NA NA NA NA NA NA NA NA NA NA ...
```

Reply:
What we need to clarify here isn't really Rle or not, but whether it is easier (code and maintenance wise) and more/less efficient (repeated only the fly generation vs somewhat larger data) to have the initialise the `DataFrame`. Let's sleep over it and touch back base next week.

Reply:
In the meantime I started to like the `Rle` idea. For some columns like `fromFile` it could make sense because it reduces the size. The accessors would *only* have to call `as.vector(Rle(1L, 1e6))`. To check for the data type we could use `class(x@value)` where `x` is a `Rle`. So, performance wise it should be *doable*.

I might try to implement a `MsBackendRleDataFrame` and we can compare.

Reply:
Ok, but I don't think we will need `MsBackendDataFrame` and `MsBackendRleDataFrame` - I assume this is only for testing?

Reply:
Yes. I want to test how the `Rle`s work in the `spectraData` - especially for subsetting etc.

Reply:
The performance comparison is in issue #17 
Regarding implementation notes we should discuss in issue #12

Reply:
As said in #12, I am happy with on-the-fly construction of data in the accessors.